### PR TITLE
LibreText team Winter 2019 Blog Post

### PR DESCRIPTION
**Pull request by lux12337**
Wrote a blog post talking about our achievements for winter quarter.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+
+\.DS_Store
```
Comment (moorepants) [inline .gitignore]: This seems very specific to your computer. Please remove (or explain).

File: content/jupyter_winter2019.rst [hunk 2]
```diff
@@ -0,0 +1,66 @@
+Creating Linux Servers for JupyterHub
+=====================================
+
+:date: 2019-04-08 00:00:00
+:tags: oer, education, jupyter, textbooks, engineering, libretexts
+:category:
```
Comment (moorepants) [inline content/jupyter_winter2019.rst]: make this "education"

File: content/jupyter_winter2019.rst [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+:slug: jupyter_winter2019
```
Comment (moorepants) [inline content/jupyter_winter2019.rst]: use hyphens instead of underscores for urls: jupyter-winter-2019

File: content/jupyter_winter2019.rst [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+:authors: Celine Liang, Kevin Krausse, Xin Luigi Chen, Xiaochen Zeng
+:summary: Blog post on setting up JupyterHub for a future computing cluster
+
+.. contents::
```
Comment (moorepants) [inline content/jupyter_winter2019.rst]: Not really needed for short posts.

File: content/jupyter_winter2019.rst [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+Background
+^^^^^^^^^^
+
+As part of the `$5M grant`_ awarded to the LibreTexts project last year,
+our team was created with two goals: to integrate Jupyter into the LibreTexts
+website and to create a computing cluster running JupyterHub to serve UC Davis.
```
Comment (moorepants) [inline content/jupyter_winter2019.rst]: to serve LibreTexts and UC Davis users.

File: content/jupyter_winter2019.rst [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+This quarter, we focused on researching how to create the cluster through
+building test servers.
+
+.. _$5M grant: https://mechmotum.github.io/blog/libretexts-grant.html
```
Comment (moorepants) [inline content/jupyter_winter2019.rst]: Links to internal pages can be done as explained here: http://docs.getpelican.com/en/3.6.3/content.html#linking-to-internal-content

**Comment by moorepants** [comment]
Thanks for the PR.

The post needs some work. It reads like a story about your work with no technical information that would be useful to a reader. I would really like to hear more about the last section that explains what you've done to improve the ansible script, what problems you hit, how you solved them, etc. There can also be some more about successes, problems, and details of running this on VMs versus the servers.

File: .gitignore (hunk 1 above)
Comment (lux12337) [reply .gitignore]: The .DS_Store file is a specific file used by MacBooks to store custom attributes for the folder it is found in. It's not a useful file for any computer other than mine, and the file is autogenerated so it is always present in the folder. Should i delete the line from .gitignore?

Comment (celine168) [reply .gitignore]: Yes, it seems like a good idea to delete it.

File: content/jupyter_winter2019.rst (hunk 2 above)
Comment (lux12337) [reply content/jupyter_winter2019.rst]: I added the "education" to category

File: content/jupyter_winter2019.rst (hunk 3 above)
Comment (lux12337) [reply content/jupyter_winter2019.rst]: fixed

File: content/jupyter_winter2019.rst (hunk 4 above)
Comment (lux12337) [reply content/jupyter_winter2019.rst]: fixed

File: content/jupyter_winter2019.rst (hunk 5 above)
Comment (lux12337) [reply content/jupyter_winter2019.rst]: fixed

File: content/jupyter_winter2019.rst (hunk 6 above)
Comment (lux12337) [reply content/jupyter_winter2019.rst]: fixed

File: .gitignore (hunk 1 above)
Comment (lux12337) [reply .gitignore]: I can't delete the file or remove them from the folder, if i delete this line from the .gitignore they will be pushed to the repo.

Comment (moorepants) [reply .gitignore]: git rm the file.

**Comment by moorepants** [comment]
What is the status on the blog post?

**Comment by celine168** [comment]
We have been prioritizing interviews and creating user stories in preparation for Yuvi's meeting this week. The blog post is on our list of objectives for next week.

**Comment by moorepants** [comment]
Ok.

File: .gitignore (hunk 1 above)
Comment (lux12337) [reply .gitignore]: removed .gitignore

**Comment by lux12337** [comment]
Hi Jason, we have added more technical details, could you please take a look at the changes and let us know if we are moving in the right direction? @moorepants

File: content/jupyter_winter2019.rst [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+ability to use and manage "dynamic partitions". When using LVM "partitions",
+known just as logical volumes, we can manage them very easily through the command
+line if we wanted to either create additional partitions, or resize/delete any
+existing partitions.
+
+While installing Ubuntu Live Server 18.04 with RAID1, we ran into an issue where
+the server failed to start. After some researching, we learned that `a bug
+<https://bugs.launchpad.net/subiquity/+bug/1785332>`__ from the Ubuntu Live Server image
+caused the installer to fail to mount the boot partition. We instead used a 
+non-live server image to successfully install Ubuntu Server 18.04 with RAID1 on the
```
Comment (moorepants) [inline content/jupyter_winter2019.rst]: What is a "non-live" server image?

Comment (moorepants) [reply content/jupyter_winter2019.rst]: Where would one find this?

File: content/jupyter_winter2019.rst [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+the server failed to start. After some researching, we learned that `a bug
+<https://bugs.launchpad.net/subiquity/+bug/1785332>`__ from the Ubuntu Live Server image
+caused the installer to fail to mount the boot partition. We instead used a 
+non-live server image to successfully install Ubuntu Server 18.04 with RAID1 on the 
+virtual machines.
+
+When installing Ubuntu Server with RAID1 and LVM on our virtual machines, we did not allot
+enough space on our hard disks for the operating system and JupyterHub combined. We determined
+that in total, the operating system and JupyterHub required about 15 GB of storage. To be safe,
+we now recommend to create two 20 GB virtual hard disks for setting up JupyterHub.
```
Comment (moorepants) [inline content/jupyter_winter2019.rst]: Good recommendation!

File: content/jupyter_winter2019.rst [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+We plan to have a stack of Ubuntu 18, RAID1, and LVM as our standard setup for each node in
+the cluster.
+
+JupyterHub Bare-Metal
+^^^^^^^^^^^^^^^^^^^^^
+
+Our next step was trying to setup a bare-metal verion of JupyterHub in our virtual machines. 
+We followed the instructions provided in the repository, `jupyterhub-deploy-teaching
+<https://github.com/mechmotum/jupyterhub-deploy-teaching>`__, to install JupyterHub on 
+our virtual machines and connect to it through the browser.
```
Comment (moorepants) [inline content/jupyter_winter2019.rst]: Mention that this is a "light fork" of the original repo for UCD uses.

File: content/jupyter_winter2019.rst [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+We ran into a few issues during the installation process.
+The Ansible script in the repository was missing some required installations.
+The package `python3-distutils` is required by JupyterHub but was not installed. We suspect that the 
+package may have been part of Ubuntu 16.04, so it is possible that the Ansible script did not need to
+specify installing `python3-disutils` previously. This was fixed in the Ansible Playbook via
```
Comment (moorepants) [inline content/jupyter_winter2019.rst]: This is true, not suspect.

File: content/jupyter_winter2019.rst [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+We ran into a few issues during the installation process.
+The Ansible script in the repository was missing some required installations.
+The package `python3-distutils` is required by JupyterHub but was not installed. We suspect that the 
+package may have been part of Ubuntu 16.04, so it is possible that the Ansible script did not need to
+specify installing `python3-disutils` previously. This was fixed in the Ansible Playbook via 
+`this commit <https://github.com/mechmotum/jupyterhub-deploy-teaching/commit/51b070a9ae3223d1919ec56323411ef455d642e5>`__.
+
+We also encountered Conda errors while installing JupyterHub. We suspect that this is 
+due to the conda submodules, which are fixed by running their updates in our automatic configuration
```
Comment (moorepants) [inline content/jupyter_winter2019.rst]: conda does not have submodules. I think you are referring to the conda-ansible repo which is included as a git submodule in our repo. clarify wording

File: content/jupyter_winter2019.rst [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+We ran into a few issues during the installation process.
+The Ansible script in the repository was missing some required installations.
+The package `python3-distutils` is required by JupyterHub but was not installed. We suspect that the 
+package may have been part of Ubuntu 16.04, so it is possible that the Ansible script did not need to
+specify installing `python3-disutils` previously. This was fixed in the Ansible Playbook via 
+`this commit <https://github.com/mechmotum/jupyterhub-deploy-teaching/commit/51b070a9ae3223d1919ec56323411ef455d642e5>`__.
+
+We also encountered Conda errors while installing JupyterHub. We suspect that this is 
+due to the conda submodules, which are fixed by running their updates in our automatic configuration 
+and deploying script, `setup.sh`.
```
Comment (moorepants) [inline content/jupyter_winter2019.rst]: Tell us what setup.sh does and why it is needed. This is a major new addition.

**Comment by moorepants** [comment]
Yes, looking better. I've left some comments.

File: content/jupyter_winter2019.rst (hunk 7 above)
Comment (celine168) [reply content/jupyter_winter2019.rst]: The non-live server image is from an alternate Ubuntu 18.04.02 Server installer. I clarified this in the blog post; fixed.

File: content/jupyter_winter2019.rst (hunk 9 above)
Comment (celine168) [reply content/jupyter_winter2019.rst]: Clarified it was a light fork, fixed.

File: content/jupyter_winter2019.rst (hunk 10 above)
Comment (celine168) [reply content/jupyter_winter2019.rst]: More decisive wording used; fixed.

File: content/jupyter_winter2019.rst (hunk 12 above)
Comment (celine168) [reply content/jupyter_winter2019.rst]: I added the configuration steps setup.sh takes and the benefits of automating the configuration and deployment process.

File: content/jupyter_winter2019.rst (hunk 8 above)
Comment (celine168) [reply content/jupyter_winter2019.rst]: Thank you!

File: content/jupyter_winter2019.rst (hunk 11 above)
Comment (celine168) [reply content/jupyter_winter2019.rst]: I clarified that ansible-conda was a repository.

**Comment by celine168** [comment]
I responded to all of the comments under the review. @lux12337 not sure if there is more to add to the blog post. Most of the other progress we have made was accomplished this quarter.

**Comment by moorepants** [comment]
Thanks for the updates. This looks good now. Merging!

**Pull request merged**